src/arithmetic/canonical.cc

```diff
@@ -765,5 +781,82 @@ Expr Simplify(Expr a, Map<Var, Range> vrange) {
 Stmt Simplify(Stmt a, Map<Var, Range> vrange) {
   return Simplify_(a, vrange);
 }
+
+Expr SimplifyCombiner(const Expr& expr, bool prune_unused_components) {
+  const Reduce* op = expr.as<Reduce>();
+  if (!op)
+    return expr;
+
+  // First simplify the results
+  Array<Expr> simplified_result;
+  for (const auto& res : op->combiner->result)
+    simplified_result.push_back(Simplify(res));
+
+  // Which components to keep
+  std::vector<int> used(op->combiner->result.size(), false);
+
+  if (prune_unused_components) {
+    // This function recursively marks the used components starting from
+    // the index idx
+    std::function<void(int)> mark_used;
+    mark_used = [&used, &simplified_result, op, &mark_used](size_t idx) {
+      // if the idx-th component was mark as used before, do nothing
+      if (used[idx]) return;
+      used[idx] = true;
+
+      // check if the idx-th result expr uses some lhs or rhs variables
+      // and recursively mark the corresponding components
+      for (size_t i = 0; i < simplified_result.size(); ++i)
+        if (!used[i]) {
+          if (ExprUseVar(simplified_result[idx], op->combiner->lhs[i]) ||
+              ExprUseVar(simplified_result[idx], op->combiner->rhs[i]))
+            mark_used(i);
+        }
+    };
+
+    // mark all used components starting from the value_index
+    mark_used(op->value_index);
+  } else {
+    // if pruning was not requested, keep all components
+    used.assign(used.size(), true);
+  }
+
+  int new_value_index = op->value_index;
+  Array<Expr> new_result;
+  Array<Expr> new_identity;
+  Array<Var> new_lhs;
+  Array<Var> new_rhs;
+  Array<Expr> new_source;
+
+  // new stuff is old stuff which is used
+  for (size_t i = 0; i < used.size(); ++i) {
+    if (used[i]) {
+      // We simplify the result and identity, but not the source
+      new_result.push_back(simplified_result[i]);
+      new_identity.push_back(Simplify(op->combiner->identity_element[i]));
+      new_lhs.push_back(op->combiner->lhs[i]);
+      new_rhs.push_back(op->combiner->rhs[i]);
+      new_source.push_back(op->source[i]);
+    } else if (static_cast<int>(i) < op->value_index) {
+      // value_index should also be adjusted
+      new_value_index--;
+    }
+  }
+
+  CommReducer new_combiner = CommReducerNode::make(new_lhs, new_rhs, new_result, new_identity);
+  return Reduce::make(new_combiner, new_source, op->axis, op->condition, new_value_index);
+}
+
+Expr RemoveEmptyReduction(const Expr& e) {
+  const Reduce* r = e.as<Reduce>();
+  if (r && r->axis.empty()) {
+    // Note that here we assume that the identity element is indeed identity.
```

Comment:
Without this assumption we would have to unroll a single iteration of reduction, i.e. instead of `r->source[r->value_index]` we would write `(*r->combiner.get())(r->combiner->identity_element, r->source)[r->value_index]`. The latter is more difficult to simplify, for example, for `max` the smallest representable number is used as the identity, and the simplifier doesn't seem to know that it is an identity. I'll add something along the lines to the comment.